Feat/#311 앱스토어의 앱 버전과 기존 앱의 버전을 추적하여 강제 업데이트 기능

isakatty opened the pull request:
## 작업내용
기존 앱 버전과 앱스토어에 올라간 앱의 major 버전을 비교하여 앱스토어를 열어주는 alert 띄우기

**⚠️ 앱스토어 연결을 위해 시뮬이 아니라 실기기 빌드 해주세요!**

| 프로젝트 info plist 수정 | 실기기 연결 | 실행화면 |
|--------|--------|--------|
| <img width="887" alt="스크린샷 2024-05-26 오후 3 00 44" src="https://github.com/Pepsi-Club/WhereMyBus-iOS/assets/133845468/5c81315f-2be9-4b5f-a9c0-eb086dc3083d"> | <img width="347" alt="스크린샷 2024-05-26 오후 3 01 24" src="https://github.com/Pepsi-Club/WhereMyBus-iOS/assets/133845468/26333fc8-030c-4ed8-8b0d-50849ea50474"> | 아래 동영상 첨부 |

https://github.com/Pepsi-Club/WhereMyBus-iOS/assets/133845468/ae0e20ca-4942-4e0b-be5d-33a261bee5db


## 리뷰요청
- @yuhaeun-la 
- @gnksbm 

## 관련 이슈
close #311

isakatty committed:
[Add] #311 App store에 등록된 app id 추가

isakatty committed:
[Add] #311 앱스토어에 등록된 앱의 최신 버전을 가져오는 함수

isakatty committed:
[Add] #311 앱스토어에 연결하는 합수

isakatty committed:
[Feat] #311 앱스토어 버전과 프로젝트 버전 비교해서 업데이트 alert 주는 함수

yuhaeun-la commented:
지수님 제가 오늘부터 목요일까지 맥북을 사용할 수 있는 환경이 아니어서, 괜찮으시면 리뷰어 분을 한명 더 등록하시는건 어떨까요? 저는 목요일 이후에 리뷰가 가능할거같아요

isakatty commented:
> 지수님 제가 오늘부터 목요일까지 맥북을 사용할 수 있는 환경이 아니어서, 괜찮으시면 리뷰어 분을 한명 더 등록하시는건 어떨까요? 저는 목요일 이후에 리뷰가 가능할거같아요

앗 ! 하은님 괜찮으시면 그렇게 하겠습니다 !

yuhaeun-la commented on Projects/App/Sources/SceneDelegate.swift:
```diff
@@ -66,5 +69,48 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+                let splitMarketingVersion = marketingVersion.split(separator: ".").map { $0 }
+                let splitCurrentProjectVersion = currentProjectVersion.split(separator: ".").map { $0 }
+                
+                if splitCurrentProjectVersion.count > 0 && splitMarketingVersion.count > 0 {
```
이 구문이 필수적으로 필요한지 궁금합니다!

isakatty replied:
82, 83은 [1.2.3] 이렇게 오는걸 .을 기준으로 나누기 위해서 필요하고,
85번째 줄은 .을 기준으로 잘 들어왔는지 확인하기 위해서 count로 조건을 준거였어요 !

yuhaeun-la commented:
> > 지수님 제가 오늘부터 목요일까지 맥북을 사용할 수 있는 환경이 아니어서, 괜찮으시면 리뷰어 분을 한명 더 등록하시는건 어떨까요? 저는 목요일 이후에 리뷰가 가능할거같아요
> 
> 앗 ! 하은님 괜찮으시면 그렇게 하겠습니다 !

네 그게 좋을 것 같습니다 ! 현재 깃헙만 둘러볼 수 있는 상황이어서 ㅠㅠ 코드 리뷰만 남기는 걸로 하겠습니다

isakatty commented on Projects/App/Sources/AppStoreCheck.swift:
```diff
@@ -0,0 +1,62 @@
+    = "itms-apps://itunes.apple.com/app/apple-store/"
+    
+    /// 앱스토어에 등록된 최신 버전 가져오는 함수
+    static public func latestVersion(completion: @escaping (String?) -> Void) {
```
아아 app delegate 에서 쓰일 코드가 아닌데 async-await 해놨던걸 거기서 쓰다가 문제가 생겨서 completion으로 변경하고 scene delegate를 하다가 다시 수정한다는걸 잊어버렸어요 🫠 
async - await으로 수정하겠습니다 !

isakatty committed:
[Chore] #311 completion block -> async await으로 수정 및 관련 코드 수정

isakatty committed:
[Remove] #311 legacy 주석 삭제

gnksbm reviewed (commented):
추가로 ignore 파일 업데이트가 안되어 있어 디버깅을 못하고있어요
올려주시면 땡겨서 추가 확인해보겠습니다~

isakatty commented on Projects/App/Sources/AppStoreCheck.swift:
```diff
@@ -0,0 +1,64 @@
+
+public final class AppStoreCheck {
+    
+    /// 프로젝트 버전
+    let appVersion = String.getCurrentVersion()
```
sceneDelegate에서 가져다 쓰려고 뒀는데 scene에서도 또 변수 만들어서 썼네요 🥲 수정하겠습니다 ~

isakatty commented on Plugins/EnvironmentPlugin/ProjectDescriptionHelpers/InfoPlist.swift:
```diff
@@ -50,7 +50,8 @@ public extension [String: Plist.Value] {
         "LOCATION_PRIVACY_URL": "$(LOCATION_PRIVACY_URL)",
-        "INQURY_URL": "$(INQURY_URL)"
+        "INQURY_URL": "$(INQURY_URL)",
+        "APPLE_ID": "$(APPLE_ID)",
```
변경하겠습니다 !

isakatty committed:
[Chore] #311 변수명 변경 및 린트 오류 수정, 사용하지 않는 변수 삭제

isakatty committed:
[Chore] Single 형태로 변경

- 한 번 구독되고 값이 방출되면 stream이 마무리되는 single 선택
- error 케이스 생성
- private init() 을 통해 객체 생성 방지

isakatty committed:
[Feat] AppStore에 올라간 정보를 받는 Endpoint 정의

- 린트에 맞춰 수정
- 싱글톤 형태로 변경

isakatty committed:
[Add] #311 AppInfoResponse, DTO 추가

isakatty committed:
[Add] #311 AppStoreRepository 생성

isakatty committed:
[Remove] #311 레거시 코드 삭제

isakatty committed:
[Refactor] #311 Endpoint Scheme 형태 변경 및 앱스토어 링크를 Endpoint 추가

isakatty committed:
[Add] #311 네트워크 통신 Single 반환 형태 추가

isakatty committed:
[Add] #311 itunes search api DTO 및 Response 모델 추가

isakatty committed:
[Refactor] #311 사용자 앱 버전 반환 형태 변경

isakatty committed:
[Fix] #311 Lint 형태 맞춰 수정

isakatty committed:
[Add] #311 버전 확인을 위한 Repository 추가

isakatty committed:
[Add] #311 Version 확인을 위한 UseCase 추가

isakatty committed:
[Add] #311 DIContainer에 UseCase 의존성 추가

isakatty committed:
[Refcator] #311 구조에 맞춰 버전 확인 적용

isakatty committed:
Merge branch 'dev' of https://github.com/Pepsi-Club/WhereMyBus-iOS into feat/#311
"[Fix] #311 dev를 받아오면서 생기는 오류 및 관련 수정 적용"

isakatty committed:
[Refactor] #311 dev에 맞춰 Network 수정

isakatty committed:
[Refactor] #311 오탈자 수정

MUKER-WON commented on Projects/Domain/Sources/UseCase/DefaultVersionCheckUseCase.swift:
```diff
@@ -0,0 +1,66 @@
+    private func needsToUpdate(_ version: AppVersionInfoResponse?) -> Bool {
+        guard let version else { return false }
+        print(version.major, getUserVersion().major)
+        return version.major > getUserVersion().major ? true : false
```
앱의 major버전만 비교해서 업데이트 유무를 판단하게 되는거면
강제 업데이트 기준으로 major버전을 올리게 되는 형태가 될거 같은데
의도하신 부분일까요??

따라서 이번 업데이트는 1.x.x 버전에서 2.x.x 버전으로 올려야 할거 같은데
강제 업데이트는 용이하지만 버전 번호가 빠르게 증가할 확률도 높아질 수도 있겠네요.
아니면 최소 요구버전 을 명시적으로 넣는 방법도 있을 수 있겠네요
의견이 궁금해요!

isakatty replied:
앱 강제 업데이트 기능에 대해 이야기가 나오던 시점에는 "major의 변화는 무조건 업데이트하게 유도하자." 이었습니다!
그래서 major를 기준으로 비교하는게 의도한 부분이 맞습니다. 그렇지만 말씀해주신대로 버전 번호가 빠르게 증가하게 된다는 점도 있고, 앱 버전 관리 규칙과 어긋나는 부분도 생기게 되네요! 
major만을 기준으로 비교할 게 아니라 마이너와 함께 비교해서 유저의 버전과 일정 이상 떨어지게 된다면 업데이트를 유도하는 형태도 괜찮을 것 같습니다. 

그리고 어제 건섭님과 이야기하며 나왔던 방식중에는, 서버와 이야기를 해서 서버에 최소 요구 버전을 두고 그 기준으로 사용자의 앱 버전을 비교해서 업데이트하는 방식도 있었습니다! 서버 측에서 작업해줄 수 있으면 이 방식이 더 좋아보이는데 이야기 해보겠습니다!
혹시 어렵게 된다면 major, minor 를 비교했을 때 2,3 버전 차이나면 업데이트 하는 방식은 어떨까요?

isakatty committed:
[Add] #311 최소버전에 대한 Endpoint 및 DTO 모델

isakatty committed:
[Refactor] #311 변경된 Endpoint에 맞춰 수정

isakatty committed:
[Add] #311 plist 수정

isakatty committed:
[Refactor] #311 버전관리 의존성 주입 주석처리

* 추후 변경 예정

isakatty committed:
[Refactor] #311 모델 리턴 타입 변경, Comparable 채택

- 강제 업데이트를 위한 정보를 담은 모델 생성
- major, minor, patch 버전 비교를 위한 Comparable 프로토콜 채택
- DTO -> Response 변환 과정 중 리턴 형태 수정

isakatty committed:
[Refactor] #311 버전 관련 데이터 소스들을 관리하는 Repository 수정

isakatty committed:
[Add] #311 버전을 비교하는 Service 객체 생성

- 추후 앱스토어 버전을 비교하는 형태 추가 예정

isakatty committed:
[Refactor] #311 버전을 비교하는 로직 수정

isakatty committed:
[Refactor] #311 최소 요구 버전 객체 네이밍 변경
